tik/tikSwap/tikSwap.cpp

```diff
@@ -88,196 +77,317 @@ int main(int argc, char *argv[])
+                if (succ->hasNPredecessors(1))
+                {
+                    // for each inst in the now predecessor-less block
+                    for (auto &it : *succ)
+                    {
+                        for (auto use : it.users())
+                        {
+                            if (auto useInst = dyn_cast<Instruction>(use))
+                            {
+                                if (!useInst->isTerminator())
+                                {
+                                    // if this unresolved use is not a terminator and has no uses, it can be trivially deleted
+                                    if (useInst->hasNUses(0))
+                                    {
+                                        badInst.insert(useInst);
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
         ind->eraseFromParent();
     }
+    for (auto inst : badInst)
+    {
+        inst->eraseFromParent();
+    }
+
+    // now verify that phis only contain valid entries
+    for (auto &fi : *sourceBitcode)
+    {
+        for (auto &bi : fi)
+        {
+            // very weird phenomenon here, but essentially we have to collect all phi entries to delete first, then delete them
+            // when all phi entries are deleted, llvm deletes the phi, so to avoid segfault we can't remove until we're done with the block
+            vector<pair<PHINode *, unsigned int>> remInd;
+            for (auto it = bi.begin(); it != bi.end(); it++)
+            {
+                if (it.getNodePtr() != nullptr)
+                {
+                    if (auto phi = dyn_cast<PHINode>(it))
+                    {
+                        // each index to remove has to be the index that results after all previous entries to remove are out
+                        unsigned int j = 0;
+                        for (unsigned int i = 0; i < phi->getNumIncomingValues(); i++)
+                        {
+                            if (find(predecessors(phi->getParent()).begin(), predecessors(phi->getParent()).end(), phi->getIncomingBlock(i)) == predecessors(phi->getParent()).end())
+                            {
+                                remInd.emplace_back(pair(phi, j));
+                            }
+                            else
+                            {
+                                j++;
+                            }
+                        }
+                    }
+                }
+            }
+            for (auto ind : remInd)
+            {
+                ind.first->removeIncomingValue(ind.second);
+            }
+        }
+    }
 
-    //verify the module
+    // verify the module
     std::string str;
     llvm::raw_string_ostream rso(str);
-    bool broken = verifyModule(*base, &rso);
+    bool broken = verifyModule(*sourceBitcode, &rso);
```

Comment:
Is this code that would be best moved to AtlasUtil?

We seem to be checking for broken modules with increasing frequency.

Reply:
I'm alright with that. want me to do it?

Reply:
Yes please.